### PR TITLE
Make feature macro undefs conditional on `-cl-ext` input

### PR DESCRIPTION
**Pull request by AGindinson**
Commits d6563b0, de262c9 and d5a2638 have provided a resolution for `opencl-c-base.h` defining OCL C 3.0 feature macros unconditionally, which interfered with Compute Runtime's platform-dependent definitions of extension lists.

An issue with this approach is that `opencl-c-base.h` provides additional defines based on `__opencl_c_atomic_scope_all_devices` being set:
```
#if defined(__opencl_c_atomic_scope_all_devices)
  memory_scope_all_svm_devices = __OPENCL_MEMORY_SCOPE_ALL_SVM_DEVICES,
#if (__OPENCL_C_VERSION__ >= CL_VERSION_3_0 || __OPENCL_CPP_VERSION__ >= 202100)
  memory_scope_all_devices = memory_scope_all_svm_devices,
#endif // (__OPENCL_C_VERSION__ >= CL_VERSION_3_0 || __OPENCL_CPP_VERSION__ >= 202100)
#endif // defined(__opencl_c_atomic_scope_all_devices)
```
(from https://github.com/llvm/llvm-project/blob/release/16.x/clang/lib/Headers/opencl-c-base.h#L387-L392)

Simply erasing the macro definition from `opencl-c-base.h` leads to memory scope values' erasure, as the command line definitions aren't available at that moment due to headers' inclusion order. Original commit d6563b0 did mention an alternative approach, which accounts for the above issue:
> Once https://reviews.llvm.org/D141297 (LLVM 16) is integrated,
  an alternative approach could be to pass `-D__undef_<feature_macro>`
  <...> at the OCL Clang's level (upon detecting the lack of
  extension-defining inputs to `-cl-ext`).

Implement just that in scope of the current commit. Said LLORG patch simply needs to be applied for CClang 14-15 instead of the original "unconditional undef" one.

### CODE REVIEW

**Comment by AGindinson** [comment]
@hewj03, @haonanya, @MateuszJablonski, could you please take a look?

File: options_compile.cpp [hunk 1]
```diff
@@ -282,6 +282,35 @@ std::string EffectiveOptionsFilter::processOptions(const OpenCLArgList &args,
...
+    // Check early exit conditions, i.e. if we can consider the language
+    // feature enabled and allow for this particular macro definition.
+    // Case 1: extension is explicitly mentioned and enabled.
+    if (extIt != extMapEnd && extIt->second)
+      continue;
+    // Case 2: the extension isn't mentioned explicitly, however "-cl-ext=+all"
+    // is present.
+    if (extIt == extMapEnd && extMapAllEnabled)
+      continue;
+    effectiveArgs.push_back(OclCFeature.insert(0, "__undef_"));
```
Comment (AGindinson) [inline options_compile.cpp:311]:
```suggestion
    effectiveArgs.push_back(OclCFeature.insert(0, "-D__undef_"));
```

**Review by hewj03 — APPROVED** [review]
LGTM.

**Comment by hewj03** [comment]
would you also add the patch to 15/16 branches?

**Comment by AGindinson** [comment]
> would you also add the patch to 15/16 branches?

Sure, I'll be creating porting CPs soon. For 16, we'd just need the `compile_options.cpp` changes.

**Comment by AGindinson** [comment]
Porting PRs: #434, #435, #436

**Pull request merged**